US59615 - Fix bgp nodeutils on eXR for new attributes (cisco_bgp_af)

brturn opened the pull request:
## TODO:
- Move ios_xr? & nexus? to Node.rb
- default_metric (need to mark as not applicable to XR and add negative tests)
- inject_map (need to mark as not applicable to XR and add negative tests)
- suppress_inactive (need to mark as not applicable to XR and add negative tests)
- test create/delete AFs (call BGP.afs and check the value?)
- distance  (need to add XR CLI)
- table_map (need to add XR CLI)
- test_route_target
- test_dampening - skipping defaults-setting on XR due to no default values on XR?
- deleted code covered by the missing_method() metafunction
- added replies_to? metafunction

Commit:
Added new features for XR

Commit:
Cleanup comments in tests.

brturn commented:
@glennmatthews @chburket Everything from the story is done, added in most of the comments from the last story as well, and deleted code which is covered by missing_method().

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -98,52 +98,34 @@ def set_args_keys(hash={}) # rubocop:disable Style/AccessorMethodName
-    def ios_xr?
-      platform == :ios_xr
+    def fail_unsupported
+      fail Cisco::UnsupportedError.new('','','')
```
I don't like this. Can we provide _any_ information for troubleshooting?

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -98,52 +98,34 @@ def set_args_keys(hash={}) # rubocop:disable Style/AccessorMethodName
-    def nexus?
-      platform == :nexus
-    end
+    # def vrf_unsupported_xr?
```
If these are no longer needed/used, please remove them rather than just commenting out.

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -98,52 +98,34 @@ def set_args_keys(hash={}) # rubocop:disable Style/AccessorMethodName
+    #   fail Cisco::UnsupportedError.new(
+    #     'bgp_af', reason, 'set',
+    #     "#{reason} is not configurable on a " \
+    #     'per-VRF basis on IOS XR') if vrf_unsupported_xr?
```
Curious why this is removed - are there now no VRF-exclusive properties left?

brturn replied:
Driving towards metafunctions; the vrf unsupported exceptions are more compactly represented in a table rather than individual functions with individual fail calls.  This will also allow merging the test matrix with the base code - previously the tests replicated the base code.  Now that I've drawn most of the data into the tests, the structure of the command tree can be queried and the appropriate result generated.  Right now, that means throwing a CliError for almost everything, but that would be easy to change.

glennmatthews commented on lib/cisco_node_utils/bgp_af.rb:
```diff
@@ -853,59 +588,16 @@ def method_missing(*args)
-    #     def default_table_map_filter
-    ##########################################
+    # Is the given name available in the YAML?
+    def responds_to?(name)
```
The relevant Ruby API is `respond_to?` not `responds_to?`. See http://technicalpickles.com/posts/using-method_missing-and-respond_to-to-create-dynamic-methods/ for a pretty good walkthrough.

brturn replied:
Reminds me that I need to write a test function for this!

glennmatthews replied:
BTW, the Puppet autogen code heavily relies on calling `respond_to?` in the NU class, so getting this working properly is a pre-requisite for getting Puppet cisco_bgp_af working again. I'm fixing this up and should have a commit ready soon.

glennmatthews commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -74,13 +74,14 @@ dampen_igp_metric:
 
 dampening:
   auto_default: false
+  multiple: true
```
Why this change?

brturn replied:
It's an error; thought 'multiple' meant 'returns a list', but that is not quite correct.

glennmatthews commented on lib/cisco_node_utils/cmd_ref/bgp_af.yaml:
```diff
@@ -252,20 +270,23 @@ route_target_import_evpn:
+    config_set_append: ' <state> table-policy <map>'
 
 table_map_filter:
+  _exclude: [cli_ios_xr]
```
This seems odd - from table_map above it looks like filter is supported on XR?
(edit) Ah, I see - filter is _not_ supported on XR, the regexp provided above in table_map is needlessly broad. OK.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -132,44 +137,41 @@ def setup
+    [:advertise_l2vpn_evpn,          :nexus,  'default', :any,                 :CliError],
+    [:advertise_l2vpn_evpn,          :nexus,  'red',     :multicast,           :CliError],
+    [:inject_map,                    :nexus,  :any,      :multicast,           :CliError],
+    [:route_target_both_auto,        :nexus,  :any,      :any,                 :CliError],
```
So the idea with here plus the above is that the route-target commands are only supported in a VRF on Nexus?

brturn replied:
Currently, it appears that "route_target_.." works only on Nexus, in a VRF, under Unicast modes.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -132,44 +137,41 @@ def setup
+    [:route_target_both_auto_evpn,   :ios_xr, :any,      :any,                 :unsupported],
+    [:route_target_import_evpn,      :ios_xr, :any,      :any,                 :unsupported],
+    [:route_target_export_evpn,      :ios_xr, :any,      :any,                 :unsupported],
+    [:inject_map,                    :ios_xr, :any,      :any,                 :unsupported],
```
What's the ordering of these entries? Can we alphabetize them or something?

brturn replied:
Will do.

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -295,16 +316,25 @@ def test_collection_empty
+      end
+    end
+
+    # config('router bgp 55',
```
delete?

brturn replied:
@chburket - You suggested I add this function; is this what you were thinking about, or did I miss the mark?

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -410,26 +425,47 @@ def test_dampening
+    # Check getters
+
+    # TODO: one returns a list of integers, the other a list of numbers
+    #       thus they don't match and the assert fails... should fix
```
Yup - this needs to be fixed. :-)

glennmatthews commented on tests/test_bgp_af.rb:
```diff
@@ -1019,35 +810,37 @@ def table_map(asn, vrf, af)
     bgp_af.table_map_set(val)
-    assert_equal(val, bgp_af.table_map,
-                 'Error: default metric value does not match set value')
+    assert_equal(val, bgp_af.table_map)
```
Why did you remove the explanatory messages in these checks?
(edit) Ah, I see - these are incorrect messages, copy-and-paste errors. Still, I'd prefer to fix them rather than removing them.

brturn replied:
Basically, all the assert messages were just typing out the same information that gets printed out when the assert fails.  This seems wasteful to me.  In some cases, the failing-assert message was incorrect and I think the original author added a specific message to correct the auto-generated messages.  Actually, it just needed the order of the parameters to be swapped to make the auto-generated message correct.  I've done this for all the asserts I've come across so far, and I prefer the auto-generated messaging since it can be internationalized more easily.

glennmatthews commented:
@brturn I'm going to merge part of these changes as part of the sync with latest develop content (which moves the route_target properties into Vrf). I'll work with you to address the merge conflicts tomorrow.

Commit:
Edits based on comments from PR

- Alphabetized method names in test matrix
- Added calls to pass method name into fail_unsupported
- Minor fixes

brturn commented:
Oops!  Forgot to run this latest one on the Nexus..  I'll start it before I leave tonight.

glennmatthews commented:
I resolved the merge conflicts and ran test_bp_af against XR and NX and it's passing - merged. Thanks for seeing this through to completion!
